Client/Frontend/Sync/BraveCore/Bookmarks/BookmarkAPIExtensions.swift

```diff
@@ -0,0 +1,255 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Data
+import BraveCore
+import BraveShared
+import CoreData
+import Shared
+
+private let log = Logger.browserLogger
+
+extension BraveBookmarksAPI {
+
+    // MARK: Internal
+    
+    // Returns the last visited folder
+    // If no folder was visited, returns the mobile bookmarks folder
+    // If the root folder was visited, returns nil
+    func lastVisitedFolder() -> BookmarkNode? {
+        guard Preferences.General.showLastVisitedBookmarksFolder.value,
+              let nodeId = Preferences.Chromium.lastBookmarksFolderNodeId.value else {
+            // Default folder is the mobile node..
+            if let mobileNode = mobileNode {
+                return mobileNode
+            }
+            return nil
+        }
+        
+        // Display root folder instead of mobile node..
+        if nodeId == -1 {
+            return nil
+        }
+        
+        // Display last visited folder..
+        if let folderNode = getNodeById(nodeId),
+           folderNode.isVisible {
+            return folderNode
+        }
+        
+        // Default folder is the mobile node..
+        if let mobileNode = mobileNode {
+            return mobileNode
+        }
+        return nil
+    }
+    
+    func lastFolderPath() -> [BookmarkNode] {
+        if Preferences.General.showLastVisitedBookmarksFolder.value,
+           let nodeId = Preferences.Chromium.lastBookmarksFolderNodeId.value,
+           var folderNode = getNodeById(nodeId),
+           folderNode.isVisible {
+            
+            // We don't ever display the root node
+            // It is the mother of all nodes
+            let rootNodeGuid = rootNode?.guid
+            
+            var nodes = [BookmarkNode]()
+            nodes.append(folderNode)
+            
+            while true {
+                if let parent = folderNode.parent, parent.isVisible, parent.guid != rootNodeGuid {
+                    nodes.append(parent)
+                    folderNode = parent
+                    continue
+                }
+                break
+            }
+            return nodes.map({ $0 }).reversed()
+        }
+        
+        // Default folder is the mobile node..
+        if let mobileNode = mobileNode {
+            return [mobileNode]
+        }
+        
+        return []
+    }
+    
+    func mobileNode() -> BookmarkNode? {
+        if let node = mobileNode {
+            return node
+        }
+        return nil
+    }
+    
+    func fetchParent(_ bookmarkItem: BookmarkNode?) -> BookmarkNode? {
```

Comment:
Is there a reason this function takes an optional bookmark?

Reply:
This method is removed because it is not used inside the application.

Fix at https://github.com/brave/brave-ios/pull/4024/commits/afa0c0001e63c7804bd48cfdf71bfc66d730557b